### PR TITLE
Metadata Submitter backend API - OpenAPI specs

### LINKED ISSUE #13: Add OpenAPI specs to project
#### Proposed solution
We should add OpenAPI specs when we have more api endpoints available. This can be added to project docs later to help with programmatic api usage. 

See more about spec here: https://swagger.io/specification/

#### DoD (Definition of Done)
- YAML file in repo specifying endpoints.

### PR DESCRIPTION
**Pull request by blankdots**
Adresses #13

### Description

View at: http://editor.swagger.io/?url=https://raw.githubusercontent.com/CSCfi/metadata-submitter/c76d7d711ed5cfdb59b494e740193c6110bfe25a/docs/specification.yml

For error messages we can also use: https://tools.ietf.org/html/rfc7807

For pagination we can use: https://tools.ietf.org/html/rfc5988 in the header the `next`, see https://developer.github.com/v3/#pagination

### Related issues

Fixes #13 

### Type of change

- [x] Documentation
- [x] New feature (non-breaking change which adds functionality)

### Testing

- [x] Tests do not apply

### Mentions

@csc-jm @otahontas what do you think about it ?

We can also extend the `schemas` endpoint where we will add new schemas or update them when we will support custom schemas.

### CODE REVIEW

**Comment by csc-jm** [comment]
/validate has nothing in the request body? 
/submit and /validate respond the same exact way?

Other than that all looks good to me

File: docs/specification.yml [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+    url: 'http://www.apache.org/licenses/LICENSE-2.0.html'
+tags:
+  - name: Submission
+    description: Endpoints used in submission
+  - name: Query
+    description: Endpoints used for data retrieval
+  - name: Manage
+    description: Endpoints used for managing data
+paths:
+  /submit:
```
Comment (otahontas) [inline docs/specification.yml:19]: Current version of `/submit` -endpoint allows insertion of multiple objects types at a time (e.g. study+sample+experiment), so response with single json accessionId doesn't reflect to what happened in db. Probably answer with json containing accessionId for each added object?

File: docs/specification.yml [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+              schema:
+                $ref: "#/components/schemas/400BadRequest"
+        401:
+          description: Unauthorized
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/401Unauthorized"
+  /validate:
+    post:
```
Comment (otahontas) [inline docs/specification.yml:50]: should have 
`
requestBody:
        content:
          multipart/form-data::
            schema:
              type: string
              format: binary
`

File: docs/specification.yml [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/401Unauthorized"
+        403:
+          description: Forbidden
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/403Forbidden"
```
Comment (otahontas) [inline docs/specification.yml:246]: 404 if nothing was found?

File: docs/specification.yml [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+      parameters:
+      - name: schema
+        in: path
+        description: Title of the XML schema.
+        schema:
+          type: string
+        required: true
+      requestBody:
+        content:
+          multipart/form-data::
```
Comment (otahontas) [inline docs/specification.yml:260]: can actually be json as well

File: docs/specification.yml [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/401Unauthorized"
+        403:
+          description: Forbidden
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/403Forbidden"
```
Comment (otahontas) [inline docs/specification.yml:319]: 404 if not found?

File: docs/specification.yml [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+        schema:
+          type: string
+        required: true
+      - name: accessionId
+        in: path
+        description: filter objects in schema using accession ID
+        schema:
+          type: string
+        required: true
+      responses:
```
Comment (otahontas) [inline docs/specification.yml:337]: patch should include json containing data that replaces old data?

File: docs/specification.yml [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+        in: path
+        description: Unique id of the targeted service.
+        schema:
+          type: string
+        required: true
+      - name: accessionId
+        in: path
+        description: filter objects in schema using accession ID
+        schema:
+          type: string
```
Comment (otahontas) [inline docs/specification.yml:377]: multipart/form-data and json for content-type?

File: docs/specification.yml [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+            application/json:
+              schema:
+                $ref: "#/components/schemas/401Unauthorized"
+  /validate:
+    post:
+      tags:
+        - Submission
+      summary: Validate XML before submission.
+      responses:
+        201:
```
Comment (otahontas) [inline docs/specification.yml:55]: Should be 200. Current response is `{"isValid: "boolean"}`

Comment (blankdots) [reply docs/specification.yml]: added that + if that is `false` there should be a `detail` key (optional) with what went wrong.

File: docs/specification.yml (hunk 6 above)
Comment (blankdots) [reply docs/specification.yml:369]: yes, it can contain just the parts that needs replacing

File: docs/specification.yml (hunk 1 above)
Comment (blankdots) [reply docs/specification.yml:19]: good catch. changed to `ObjectsCreated` so that it also returns the schemas of the objects that were added.

**Comment by blankdots** [comment]
@csc-jm and @otahontas good points addressed them in new commits, let me know of other changes.

File: docs/specification.yml (hunk 6 above)
Comment (otahontas) [reply docs/specification.yml:369]: yeah, but shouldn't we add it to api specs?

**Pull request merged**